### PR TITLE
feat: add support for using resolver object directly in settings

### LINKED ISSUE #75: [Proposal] Support using any object satisfied `Resolver` type in `import-x/resolver` options
The resolver currently supports three reference methods:

- name -> eslint-import-resolver-name
- npm package name
- filesystem path

I propose to support the direct use of any object satisfied `Resolver` type, and I further propose configuration other than this should be deprecated in the future.

Advantage:

- Reduce complexity when trying to require resolvers
- Defining a custom resolver is more convenient

For advantage 1, you can refer to ESLint’s thoughts on the introduction of flat config:

> Use native loading for everything - one of our biggest regrets about eslintrc was recreating the Node.js require resolution in a custom way. This was a significant source of complexity and, in hindsight, unnecessary. Going forward, we wanted to leverage the loading capabilities of the JavaScript runtime directly.

Reference: https://eslint.org/blog/2022/08/new-config-system-part-2/#the-goals-of-flat-config

For advantage 2, there may be cases when existing resolvers don't meet the needs and needs to be temporarily added.

In this case, if the direct use of objects is supported, the steps of "creating new file -> processing filesystem path" can be eliminated, and the object can be created in place and used immediately.

For example, suppose I want to bundle the eslint configuration and all dependencies into a single file for use. In this case, none of the above three configuration methods will work. However, if the feature of directly using the resolver object is provided, my needs can be well met.

In order to support the use of `Resolver` objects, the configuration structure of the `import-x/resolver` configuration item needs to be modified.

Before:
```ts
export type ImportResolver =
  | LiteralUnion<'node' | 'typescript' | 'webpack', string>
  | {
      node?: boolean | NodeResolverOptions
      typescript?: boolean | TsResolverOptions
      webpack?: WebpackResolverOptions
      [resolve: string]: unknown
    }
```

After:
```ts
export type Resolver = {
  interfaceVersion?: 1 | 2
  resolve: ResolverResolve
  resolveImport: ResolverResolveImport
}

export type ResolverName = LiteralUnion<
  'node' | 'typescript' | 'webpack',
  string
>

export type ResolverRecord = {
  node?: boolean | NodeResolverOptions
  typescript?: boolean | TsResolverOptions
  webpack?: WebpackResolverOptions
  [resolve: string]: unknown
}

export type ResolverObject = {
  // node, typescript, webpack...
  name: ResolverName

  // Enabled by default
  enable?: boolean

  // Options passed to the resolver
  options?:
    | NodeResolverOptions
    | TsResolverOptions
    | WebpackResolverOptions
    | unknown

  // Any object satisfied Resolver type
  resolver: Resolver
}

export type ImportResolver =
  | ResolverName
  | ResolverRecord
  | ResolverObject
  | ResolverName[]
  | ResolverRecord[]
  | ResolverObject[]
```

### PR DESCRIPTION
**Pull request by GoodbyeNJN**
close #75

### CODE REVIEW

**Comment by GoodbyeNJN** [comment]
@JounQin @SukkaW Sorry to bother you, but can anyone push this PR?
I proposed this feature before, but there was no discussion, so I decided to implement it first.
If you are free, could you review this feature proposal and code?

File: src/types.ts [hunk 1]
```diff
@@ -32,14 +32,73 @@ export type DocStyle = 'jsdoc' | 'tomdoc'
...
+  modulePath: string,
+  sourceFile: string,
+  config: unknown,
+) => ResolvedResult
+
+export type ResolverResolveImport = (
+  modulePath: string,
+  sourceFile: string,
+  config: unknown,
+) => string | undefined
```
Comment (SukkaW) [inline src/types.ts:57]: Is it possible to make the `config` generic? In case some resolver library authors want to use these types.

**Comment by SukkaW** [comment]
> @JounQin @SukkaW Sorry to bother you, but can anyone push this PR? I proposed this feature before, but there was no discussion, so I decided to implement it first. If you are free, could you review this feature proposal and code?

Sorry for the late reply. I like this change very much! I have added only a small review suggestion, and I have approved the CI to run. We can ship this as long as all existing test cases are passed.

Another thought, would you mind adding a test case for your change as well? So that we can test against the new config.

**Comment by GoodbyeNJN** [comment]
> Another thought, would you mind adding a test case for your change as well? So that we can test against the new config.

I have added test cases and reorganized the test set. Could you please approve running CI again?

**Comment by GoodbyeNJN** [comment]
One more thing, as I said in my proposal #75:

> I propose to support the direct use of any object satisfied Resolver type, and I further propose configuration other than this should be deprecated in the future.

If you agree with this, I will mark the other methods as deprecated in the documentation. If you don't agree with this, I will just add the new method to the documentation without making other changes.

File: README.md [hunk 2]
```diff
@@ -298,6 +298,40 @@ module.exports = {
 }
 ```
 
+- use the `import` or `require` syntax to directly import the resolver object:
+
+```js
+// .eslintrc.mjs
+import myResolver from '../../../my-resolver'
```
Comment (SukkaW) [inline README.md:305]: Ideally, we could use `eslint-import-resolver-typescript`, `eslint-import-resolver-vite`, and `eslint-import-resolver-oxc` as an example! What do you think?

**Review by SukkaW — APPROVED** [review]
LGTM!

File: .changeset/red-rings-pay.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"eslint-plugin-import-x": patch
```
Comment (SukkaW) [inline .changeset/red-rings-pay.md:2]:
```suggestion
"eslint-plugin-import-x": minor
```

**Pull request merged**